### PR TITLE
add image analysis to github workflow

### LINKED ISSUE #42: Track Image Size in PRs
Given this will increase in size over time, it would be important for us to be conscious about how much we are adding to the base image.

### PR DESCRIPTION
**Pull request by PiyushRaj927**
Fixes #42 
This PR add workflows that analyze the size differences in the Docker image. The key features of this PR are:
1. Automatically hides old bot messages.
2. Updates the Docker image's current statistics when commits are posted to the main branch.

The process involves two workflows: image-analysis.yml and image-analysis-comment.yml.

The image-analysis.yml workflow, which lacks write permissions on issues, is responsible for importing the image, conducting the analysis, exporting the details as comments in the pull request, and updating the comparison metrics when a push occurs on the main branch.

On the other hand, the image-analysis-comment.yml workflow writes the comments to their respective issues and hide bot messages from old runs.

### CODE REVIEW

**Comment by PiyushRaj927** [comment]
Hey @singhalkarun can you review the PR when you get the chance. The test is failing because forked repos can't have write access in PRs. You can see it working in my forked repo at https://github.com/PiyushRaj927/WarpSQL/pull/3

File: image_size.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
+async function saveMetricsToFile(metrics, fs) {
+  const filePath = "image-metrics.json";
```
Comment (jayanth-kumar-morem) [inline image_size.js:2]: Can we take this as an arg ?

File: image_size.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+  try {
+    await fs.promises.writeFile(filePath, jsonData);
+    console.log(`Metrics saved to ${filePath}`);
+  } catch (error) {
+    console.error(`Error saving metrics to ${filePath}: ${error}`);
+  }
+}
+
+async function readMetricsFromFile(fs) {
+  const filePath = "image-metrics.json";
```
Comment (jayanth-kumar-morem) [inline image_size.js:14]: Same cmt

File: image_size.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  const workspace = core.getInput("workspace", { required: true });
+
+  if (context.eventName == "pull_request") {
+    const existingMetrics = (await readMetricsFromFile(fs)) || [];
+    const metricToCompare =
+      existingMetrics[
+      existingMetrics.findIndex((metric) => metric.imageId === imageType)
+      ];
+
+    let githubMessage = `### :bar_chart: ${imageType} Image Analysis  (Commit: ${commitSHA} )
```
Comment (jayanth-kumar-morem) [inline image_size.js:143]: I suggest we have a template function which can take the args and return the githubMessage.

**Review by jayanth-kumar-morem — APPROVED** [review]
+1

**Comment by PiyushRaj927** [comment]
> Hey @singhalkarun can you review the PR when you get the chance. The test is failing because forked repos can't have write access in PRs. You can see it working in my forked repo at [PiyushRaj927#3](https://github.com/PiyushRaj927/WarpSQL/pull/3)

While I was researching more about the access available to the workflows, I discovered that the `pull_request` event for forked pull requests doesn't have write access by default, as it poses a security risk. I found a better approach, which involves a two-step process, explained in detail in https://stackoverflow.com/questions/69499645/how-to-securely-allow-github-actions-to-check-pr-and-post-results-in-comment. I will implement this method and incorporate the suggested changes mentioned in the initial review.

**Comment by singhalkarun** [comment]
> sts doesn't have writ

Hey @PiyushRaj927 can we create a different workflow for image analysis if possible to keep things separate?

**Comment by PiyushRaj927** [comment]
I have tested the workflow and made the necessary changes. You can see it working at https://github.com/anmolchhabra21/WarpSQL/pull/25.

**Comment by PiyushRaj927** [comment]
Hey @singhalkarun, I have added comments and formatted the code. If any part is unclear, please let me know, and I'll provide clarifications. I will also include a document explaining the workflow soon

**Pull request merged**